Performance optimizations for the router

pnarayanan opened the pull request:
This patch introduces two main changes: a) Adding a deterministic
wait-notify flow for the ChunkFiller thread to go to sleep and get
notified whenever there is work to be done. b) Adding a "wakeUp" for the
NetworkClient that will in turn wake up the associated Selector from its
poll if an operation manager becomes poll-eligible asynchronously to the
RequestResponseHandler thread.

Also adds couple of miscellaneous fixes.

Fixes #378 and #377 

**Coverage**
```
DeleteManager                   100% (2/ 2)     100% (10/ 10)   91.9% (68/ 74)
DeleteOperation                 100% (3/ 3)     100% (21/ 21)   92.9% (105/ 113)
GetBlobInfoOperation            100% (2/ 2)     100% (14/ 14)   91.7% (100/ 109)
GetBlobOperation                100% (7/ 7)     89.8% (44/ 49)  91.6% (230/ 251)
GetManager                      100% (2/ 2)     100% (12/ 12)   78.3% (54/ 69)
GetOperation                    100% (1/ 1)     100% (10/ 10)   100% (24/ 24)
GetRequestInfo                  100% (1/ 1)     100% (1/ 1)     100% (3/ 3)
NonBlockingRouter               100% (3/ 3)     100% (28/ 28)   86.4% (133/ 154)
NonBlockingRouterFactory        100% (1/ 1)     100% (3/ 3)     86.4% (19/ 22)
NonBlockingRouterMetrics        100% (6/ 6)     78.6% (11/ 14)  92% (103/ 112)
OperationCompleteCallback       100% (1/ 1)     100% (3/ 3)     100% (10/ 10)
PutManager                      100% (4/ 4)     100% (16/ 16)   90.3% (102/ 113)
PutOperation                    100% (7/ 7)     100% (62/ 62)   90% (288/ 320)
ReadyForPollCallback            100% (1/ 1)     100% (2/ 2)     100% (3/ 3)
```

Primary reviewers: Casey & Gopal
Estimate: ~ 30 min.

nsivabalan commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferAsyncWritableChannelTest.java:
```diff
@@ -177,6 +177,28 @@ public void useAfterCloseTest()
+            }
+          }
+        });
+    channel.write(ByteBuffer.allocate(5), null);
```
Can you check that the atomic booleans are false before calling write.

nsivabalan commented on ambry-network/src/test/java/com.github.ambry.network/NetworkClientTest.java:
```diff
@@ -461,6 +472,28 @@ public void poll(long timeoutMs, List<NetworkSend> sends)
   /**
+   * Clear the wakeup state of this NetworkClient.
+   */
+  public void clearWakeUpState() {
```
I don't see this is being used now. May I know the purpose of this method then.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -46,13 +47,20 @@
   private final NotificationSystem notificationSystem;
   private final Time time;
   private final Thread chunkFillerThread;
+  //private final AtomicInteger operationSetState = new AtomicInteger(0);
```
why commented out

pnarayanan replied:
removed.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -83,18 +91,35 @@ public void registerRequestToSend(PutOperation putOperation, RequestInfo request
+    this.chunkArrivalListener = new ByteBufferAsyncWritableChannel.ChannelEventListener() {
+      @Override
+      public void onEvent(ByteBufferAsyncWritableChannel.EventType e) {
+        synchronized (chunkFillerSynchronizer) {
```
do we need to wake up for all types of events?

pnarayanan replied:
yes - in the off chance that it gets woken up and there is nothing for it to do, it is okay.

nsivabalan commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -83,18 +91,35 @@ public void registerRequestToSend(PutOperation putOperation, RequestInfo request
+        synchronized (chunkFillerSynchronizer) {
+          // At this point, the chunk for which this notification came in (if any) could already have been consumed by
+          // the chunk filler, and this might unnecessarily wake it up from its sleep, which should be okay.
+          areChunksAvailableForFilling = true;
```
when do we set this variable to false again.

pnarayanan replied:
This doesn't seem right, will fix.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/GetBlobOperationTest.java:
```diff
@@ -538,6 +540,8 @@ private void assertSuccess(ReadableStreamChannel readableStreamChannel, CountDow
       ByteBufferAsyncWritableChannel asyncWritableChannel = new ByteBufferAsyncWritableChannel();
       long written;
       Future<Long> readIntoFuture = readableStreamChannel.readInto(asyncWritableChannel, null);
+      Assert.assertTrue("ReadyForPollCallback should have been invoked as readInto() was called",
```
minor. not mandatory though. Shouldn't we have a similar assert when a new operation is added.

nsivabalan replied:
oh, I didn't realize that mockNetworkClient don't call super.wakeUp(). Is that something we need ?

pnarayanan replied:
The test just verifies that the callback is invoked. The `MockNetworkClient` is just a helper class for that.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java:
```diff
@@ -74,9 +73,19 @@ private void setOperationParams() {
-    VerifiableProperties verifiableProperties = new VerifiableProperties((props));
     MockClusterMap mockClusterMap = new MockClusterMap();
+    try {
+      Properties props = getNonBlockingRouterProperties("NotInClusterMap");
```
minor. You can take out the creation of props and verifiable props outside of try block and over-ride it for next test case

pnarayanan replied:
done.

nsivabalan commented on ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java:
```diff
@@ -90,12 +99,11 @@ public void testNonBlockingRouterFactory()
   public void testRouterBasic()
       throws Exception {
-    Properties props = getNonBlockingRouterProperties();
+    Properties props = getNonBlockingRouterProperties("DC1");
```
can you define a constant for "DC1"

pnarayanan replied:
I think that is unnecessary and non-standard for our tests.

nsivabalan commented:
done with one pass

xiahome commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferAsyncWritableChannel.java:
```diff
@@ -31,10 +31,46 @@
+  /**
+   * Callback that can be used to listen to events that happen inside this channel.
+   */
+  public interface ChannelEventListener {
```
Just would like to confirm that even making `ChannelEventListener` and `EventType` public, they are still within `ByteBufferAsyncWritableChannel` because of the logic relationship? I.e., the interface and enum are only meaningful to `ByteBufferAsyncWritableChannel`?

vgkholla replied:
@pnarayanan I thought we discussed a generic listener. Why the change?

pnarayanan replied:
Yes - the interface and enum are `ByteBufferAsyncWritableChannel` specific, as the event types may not be generic across all `AsyncWritableChannel` implementations. An implementation could have other events and those could be of interest to the creator of of that channel (who registers the listener).

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -83,18 +91,35 @@ public void registerRequestToSend(PutOperation putOperation, RequestInfo request
+          // At this point, the chunk for which this notification came in (if any) could already have been consumed by
+          // the chunk filler, and this might unnecessarily wake it up from its sleep, which should be okay.
+          areChunksAvailableForFilling = true;
+          if (isChunkFillerThreadAsleep) {
```
I am wondering how necessary this check is. Can we just notify the `chunkfiller` when some data is ready from the async channel?

vgkholla replied:
+1

pnarayanan replied:
just avoiding a notify() when the other thread is not in a wait().

xiahome commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -259,24 +290,28 @@ void completePendingOperations() {
             if (!op.isChunkFillComplete()) {
               op.fillChunks();
-              allChunksFilled = false;
+              idleIteration = false;
```
It seems that the logic of using `idleIteration`is, because this loop there is nothing for me to do, so I will assume the next loop there will also be nothing to do, so I will go to sleep for the next loop. Do we have an assumption like this? Curious if the performance would be substantially impacted if `idleIteration` is removed.

pnarayanan replied:
if any work is done in an iteration, we do not want to attempt the sleep as there is a high likelihood that work could get done in the next iteration as well.

vgkholla commented on ambry-commons/src/main/java/com.github.ambry.commons/ByteBufferAsyncWritableChannel.java:
```diff
@@ -31,10 +31,46 @@
+   * List of events of interest to the consumer of the content in this channel.
+   */
+  public enum EventType {
+    ChunkArrival,
```
Can we call this write?

pnarayanan replied:
Done.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferAsyncWritableChannelTest.java:
```diff
@@ -177,6 +177,30 @@ public void useAfterCloseTest()
   }
 
+  @Test
+  public void testChannelEventNotification()
```
doc

pnarayanan replied:
done.

vgkholla commented on ambry-commons/src/test/java/com.github.ambry.commons/ByteBufferAsyncWritableChannelTest.java:
```diff
@@ -177,6 +177,30 @@ public void useAfterCloseTest()
+          }
+        });
+    assertFalse("No chunk arrival notification should have come in before any write", chunkArrivalNotified.get());
+    assertFalse("No close event notification should have come in before a close", closeNotified.get());
```
move this just before line 200? cc: @nsivabalan

pnarayanan replied:
done.

vgkholla commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -219,6 +218,13 @@ public void close() {
   }
 
   /**
+   * Wake up the NetworkClient if it is within a {@link #sendAndPoll(java.util.List)} sleep.
```
I think this documentation could be a little more descriptive and perhaps also have a pointer (via the @see tag) to the java nio selector wakeup doc.

pnarayanan replied:
done.

vgkholla commented on ambry-network/src/test/java/com.github.ambry.network/NetworkClientTest.java:
```diff
@@ -218,6 +218,16 @@ public void testExceptionOnPoll() {
+   * Test that the NetworkClient wakeUp wakes up the associated Selector.
+   */
+  @Test
+  public void testWakeUp() {
```
minor: the actual function is called `wakeup()` (all lower case)

pnarayanan replied:
done.

vgkholla commented on ambry-network/src/main/java/com.github.ambry.network/NetworkClient.java:
```diff
@@ -49,8 +49,7 @@
   private final int checkoutTimeoutMs;
-  // @todo: this needs to be empirically determined.
-  private final int POLL_TIMEOUT_MS = 1;
+  private final int POLL_TIMEOUT_MS = 300;
```
i don't think this should be 300. I think it should be set to the value of request timeouts and therefore should be configurable when starting up the network client.

pnarayanan replied:
This should not be configurable as the user is only concerned with request timeouts. But yes, this should be a function of the request timeout and should be smaller. Will make the `sendAndPoll()` take an explicit timeout and will pass in a value that is an order of magnitude smaller than the request timeout.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/DeleteManager.java:
```diff
@@ -102,6 +106,7 @@ void submitDeleteBlobOperation(String blobIdString, FutureResult<Void> futureRes
       DeleteOperation deleteOperation =
           new DeleteOperation(routerConfig, routerMetrics, responseHandler, blobId, futureResult, callback, time);
       deleteOperations.add(deleteOperation);
+      readyForPollCallback.onPollReady();
```
not sure why we need this. Why not just call this from the `OperationController`?

vgkholla replied:
Is it about consistency?

pnarayanan replied:
Yes. But will have this done right after operation submission in the caller code as you suggested in one of the other comments.

vgkholla replied:
This change wasn't made?

pnarayanan replied:
Yeah, seem to have missed. Will make the change.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java:
```diff
@@ -216,8 +216,8 @@ private void cleanupExpiredInflightRequests() {
         itr.remove();
-        NonBlockingRouterMetrics.NodeLevelMetrics dataNodeBasedMetrics =
-            routerMetrics.getDataNodeBasedMetrics(deleteRequestInfo.replica.getDataNodeId());
+        responseHandler
```
your change seems correct but it seems totally unrelated to what was there previously. Do you know why?

pnarayanan replied:
Two different changes. The line was removed as the variable was not getting used.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -305,6 +310,7 @@ public long getSize() {
       if (operationException.get() != null) {
         completeRead();
       }
+      readyForPollCallback.onPollReady();
```
is it necessary here?

pnarayanan replied:
Yes - it could be that the initial chunks were fetched but was not getting streamed out because the `asyncWritableChannel` was not available.

vgkholla replied:
But then why are we ready for a poll callback? you don't need to get any new chunks from the server yet.

pnarayanan replied:
the write to the channel also happens during the poll.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/GetManager.java:
```diff
@@ -123,8 +129,9 @@ void submitGetBlobOperation(String blobId, FutureResult<ReadableStreamChannel> f
-              operationCompleteCallback, blobIdFactory, time);
+              operationCompleteCallback, readyForPollCallback, blobIdFactory, time);
       getOperations.add(getBlobOperation);
+      readyForPollCallback.onPollReady();
```
I would prefer doing all the `onPollReady()` on operation arrival in `OperationController` and passing in the `readyForPollCallback` only when required (i.e. `GetBlobOperation` and `ChunkFiller`).

pnarayanan replied:
Yes, that makes sense.

xiahome replied:
I think the current way is correct. When a new operation is being submitted, the `NetworkClient` is waken up by the operation-submitting thread, which is different from the `OperationController` thread (which may be sleeping already).

vgkholla replied:
AFAIK, the `OperationController` doesn't have a thread? The `OperationController` code runs in the context of the operation-submitting thread.

In fact, all I'm recommending is that the code be moved. Not the context of thread where the code is run.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/PutManager.java:
```diff
@@ -221,6 +246,12 @@ boolean isOpen() {
     if (isOpen.compareAndSet(true, false)) {
+      synchronized (chunkFillerSynchronizer) {
+        if (isChunkFillerThreadAsleep) {
+          chunkFillerThreadSleepDisabled = true;
```
not sure why you need this new variable. Won't checking `isOpen` be enough ? (both variables seem to carry the same piece of information).

pnarayanan replied:
yeah, `isOpen` is enough.

vgkholla commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -142,12 +143,16 @@
-      ReadableStreamChannel channel, FutureResult<String> futureResult, Callback<String> callback, Time time)
+      ReadableStreamChannel channel, FutureResult<String> futureResult, Callback<String> callback,
+      ReadyForPollCallback readyForPollCallback,
+      ByteBufferAsyncWritableChannel.ChannelEventListener channelEventListener, Time time)
```
prefer naming this is as `chunkArrivalListener` since the `ReadableStreamChannel` passed in has the name `channel`.

pnarayanan replied:
done.

pnarayanan replied:
actually will name this as `writableChannelEventListener` as the listener is not just about chunk arrival but for all kinds of events.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/ChunkFillTest.java:
```diff
@@ -112,8 +112,11 @@ public void testChunkNumAndSizeCalculations()
     random.nextBytes(putUserMetadata);
```
any possible specific tests to test the sleep and wake up functionality?

pnarayanan replied:
will try to add a few more tests.

pnarayanan replied:
added a test.

cgtz commented on ambry-router/src/main/java/com.github.ambry.router/PutOperation.java:
```diff
@@ -796,7 +804,9 @@ private void cleanupExpiredInFlightRequests() {
           onErrorResponse(entry.getValue().replicaId);
-          chunkException = new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut);
+          responseHandler.onRequestResponseException(entry.getValue().replicaId,
+              new IOException("Timed out waiting for a response"));
```
Could we just pass this exception into the `RouterException` constructor in cases like this?

pnarayanan replied:
There is no "real" exception here that could be associated with the `RouterException`. The `IOException` is an exception specifically created for the `ResponseHandler` as it needs one.

cgtz replied:
Okay, I see

cgtz commented on ambry-router/src/test/java/com.github.ambry.router/ChunkFillTest.java:
```diff
@@ -194,8 +197,11 @@ private void fillChunksAndAssertSuccess()
+        new ReadyForPollCallback(networkClientFactory.getNetworkClient()), null, time);
     numChunks = op.getNumDataChunks();
     compositeBuffers = new ByteBuffer[numChunks];
     final AtomicReference<Exception> operationException = new AtomicReference<Exception>(null);
```
Do we have a test that ensures that a ChunkFiller thread shuts down when notified and PutManager is closed?

pnarayanan replied:
will see if I can add one.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/MockNetworkClient.java:
```diff
@@ -0,0 +1,44 @@
+import java.io.IOException;
+
+
+class MockNetworkClient extends NetworkClient {
```
is there no way to avoid this class? Also, needs doc

pnarayanan replied:
No. And not sure why we need to, this is exactly what is needed here to test what we want (except for having to pass in a few things in the constructor).
Added docs.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java:
```diff
@@ -74,9 +73,19 @@ private void setOperationParams() {
+      VerifiableProperties verifiableProperties = new VerifiableProperties((props));
+      router = (NonBlockingRouter) new NonBlockingRouterFactory(verifiableProperties, mockClusterMap,
+          new LoggingNotificationSystem()).getRouter();
+      Assert.fail("NonBlockingRouterFactory instantiation should have failed because the router datacenter is not in "
```
could you please keep _only_ that statement that throws the exception and the `fail()` in the `try` block?

pnarayanan replied:
done.

vgkholla commented:
Could you post unit test coverage?

cgtz commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java:
```diff
@@ -117,8 +117,10 @@ private void cleanupExpiredInFlightRequests() {
+            new IOException("Timed out waiting for a response"));
         setOperationException(
-            new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut));
+            new RouterException("Timed out waiting for a response", RouterErrorCode.OperationTimedOut));
```
Same thing here.  Should we pass the `IOException` into the `RouterException`?

pnarayanan replied:
same as above.

cgtz commented on ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java:
```diff
@@ -493,7 +499,9 @@ private void cleanupExpiredInFlightRequests() {
-          chunkException = new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut);
+          responseHandler.onRequestResponseException(entry.getValue().replicaId,
+              new IOException("Timed out waiting for a response"));
+          chunkException = new RouterException("Timed out waiting for a response", RouterErrorCode.OperationTimedOut);
```
Same thing here. Should we pass the IOException into the RouterException?

pnarayanan replied:
same as above. The `IOException` does not contain any useful information in order to be embedded into the `RouterException`.

pnarayanan commented:
Addressed comments. Fixed a race condition int the chunk filler wait-notify flow.
Will add a couple of additional tests. In the meanwhile, please continue reviewing.

pnarayanan commented:
Comments are addressed, coverage has been updated. @cgtz @vgkholla please take a look.

cgtz commented:
LGTM

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java:
```diff
@@ -532,6 +533,59 @@ public void testRouterClosingOnRequestResponseHandlerThreadException()
   }
 
+  @Test
+  public void testChunkFillerSleep()
```
docs

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java:
```diff
@@ -532,6 +533,59 @@ public void testRouterClosingOnRequestResponseHandlerThreadException()
+    Assert
+        .assertTrue("ChunkFillerThread should have gone to WAITING state as the only active operation is now complete",
+            waitForThreadState(chunkFillerThread, Thread.State.WAITING));
+    future.await();
```
timed wait.

vgkholla commented on ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java:
```diff
@@ -532,6 +533,59 @@ public void testRouterClosingOnRequestResponseHandlerThreadException()
+    assertCloseCleanup();
+  }
+
+  private boolean waitForThreadState(Thread thread, Thread.State state) {
```
docs

vgkholla commented:
LGTM. Will check out the doc commit.

pnarayanan commented:
@vgkholla you can take a look and merge if it looks good. Thanks.
